[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117168
* #117093
* #117092
* __->__ #116545

We added an extra sleep and make it configurable so that users can set an extra wait to ensure all ranks have dumped the debug info.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1064,6 +1068,10 @@ void ProcessGroupNCCL::waitForDumpOrTimeout(
   auto futStatus = fut.wait_for(std::chrono::seconds(timeout_sec));
+  // This extra sleep is needed to ensure all ranks get enough time to dump
+  // debug info when timeout.
+  std::this_thread::sleep_for(
```
it may be better to move this sleep outside of the helper, and use wait_until instead of wait_for.  Why?

1) maybe not all callsites of this function care about waiting for other threads (e.g. if dumping via pipe)
2) wait_until lets us observe how long the dumping took on this rank first, and any other operations we might want to do before we call wait, then finally see if we have elapsed enough time or we need to sleep.  And if we sleep, maybe we can sleep for a shorter time than the sleep variable, taking into account whatever time we already spent

fduwjj replied:
I understand the first point, but I am not sure if I am getting #2. Is the reason of changing from `wait_for` to `wait_until` because we want to sleep actually less? `future::wait_until` will also return if future has a value, IIUC.

wconstab replied:
yes, sleep less, if we already spent the time doing some other work.

fut = asyncDebugDump()
do somethign else (e.g. desyncdebug)
waitForFut(fut)
finally, sleep until()

this way, sleep until has 2 opportunities to sleep less.  (a) is if 'something else' took time, (b) is if waitForFut spent time.

fduwjj replied:
Actually for #1, looks like dumping via pipe does not even call this helper function..

fduwjj replied:
OK, I think I now get what you mean and changed the code accordingly.

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1491,8 +1503,11 @@ void ProcessGroupNCCL::watchdogHandler() {
+              std::chrono::milliseconds(waitTimeoutDumpInMilSec_);
           optAsyncDebugDump = launchAsyncDebugDump();
           waitForDumpOrTimeout(*optAsyncDebugDump);
+          extraWaitForDumpUntil(wakeUpTime);
```
i’m wondering if it would be better to keep 'extraWaitForDumpUntil' as a separate helper or, maybe pass the ‘wakeup_time’ as a 3rd argument to `waitForDumpOrTimeout`.  It seems like that would satisfy all the cases we have today.

separately, we could make a helper for getWakeupTime(time_interval) that we use to compute `wakeUpTime = std::chrono::steady_clock::now() +
              std::chrono::milliseconds(waitTimeoutDumpInMilSec_)`

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -738,6 +738,8 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   heartbeatTimeoutInSec_ =
       getCvarInt(TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC, 60 * 10 /*10 Mins*/);
+  waitTimeoutDumpInMilSec_ =
+      getCvarInt(TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC, 2000);
```
new proposal: we go back to your original PR (sorry, my fault!) but make it an env.  let users specify EXTRA_SLEEP_AFTER_DUMPING or something.  Make it default to 0 or 1s maybe, so unit tests are fast.  Set it to 10 (or 2s) or something for gen-ai trainers for now, if they are ok with that

fduwjj replied:
Agree, this PR is not a long-term fix. It just fixes case2 and when all ranks share pretty about the similar amount of time (small variance) indeed. And yes, if there is a supervisor, there is no need to have this heuristic sleep any more.

wconstab replied:
yea, i think having a configurable, 'sleep_for()' is best then.  We can avoid slowing down exit for most users, increase chances of dump for gen-ai, and still let them override to sleep longer if they are in some debug-hell (or shorter, if they are at the point 1-2 sec matters)

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab reviewed (changes requested):
LGTM. but one minor nit- can you add a doc-comment at the place we define the ENV explaining the usage of the env?  we should be doing that for every ENV.  (we can add for other ENVs in other PR but just cover this one here)

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@wconstab sure.

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -768,6 +768,8 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   monitorThreadEnabled_.store(getCvarBool(TORCH_NCCL_ENABLE_MONITORING, true));
   heartbeatTimeoutInSec_ =
       getCvarInt(TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC, 60 * 10 /*10 Mins*/);
+  waitTimeoutDumpInMilSec_ =
```
Doc for new env

fduwjj replied:
Wait, I think the tradition here is to have the doc for this member in the hpp file?

wconstab replied:
Yea that's fine too

wconstab reviewed (approved):
Lgtm, except for all 3 PR let's do the end to end test before landing any

fduwjj commented:
@wconstab Sure, I am testing the stack now. And two more PRs on top to go: https://github.com/pytorch/pytorch/pull/117092. Thanks!

fduwjj commented:
@pytorchbot merge -i

fduwjj committed:
Update on "[c10d] Add extra sleep in waitForDumpOrTimeout to ensure enough time for all ranks dump debug info"


We added an extra sleep and make it configurable so that users can set an extra wait to ensure all ranks have dumped the debug info.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
